docs/using-the-compiler.rst

```diff
@@ -0,0 +1,262 @@
+              runtimeLinkReferences: {
+              }
+          },
+          functionHashes:
+          metadata: // see below
+          ewasm: {
+              wast: // S-expression format
+              wasm: //
+          },
+          userdoc: // Obsolete
+          devdoc: // Obsolete
+          natspec: // Combined dev+userdoc
+        }
+      },
+      formal: {
+        "why3": "..."
+      },
+      sourceList: ["source1.sol", "source2.sol"], // this is important for source references both in the ast as well as in the srcmap in the contract
+      sources: {
+        "source1.sol": {
+          "AST": { ... }
+        }
+      }
+    }
+
+Metadata Output
+---------------
+
+Note that the actual bytecode is not part of the metadata because the hash
+of the metadata structure will be included in the bytecode itself.
+
+This requires the compiler to be able to compute the hash of its own binary,
+which requires it to be statically linked. The hash of the binary is not
+too important. It is much more important to have the commit hash because
+that can be used to query a location of the binary (and whether the version is
+"official") at a registry contract.
+
+    {
+      // The version of the metadata format (required field)
+      version: "1",
+      // Required field
+      language: "Solidity",
+      // Required field, the contents are specific to the language
+      compiler: {
+        name: "solc",
+        version: "0.4.5-nightly.2016.11.15+commit.c1b1efaf.Emscripten.clang",
+        // Optional hash of the compiler binary which produced this output
+        keccak256: "0x123..."
+      },
+      // This is a subset of the regular compiler input
+      sources:
+      {
+        "myFile.sol": {
+          "keccak256": "0x123...",
+          "url": "bzzr://0x56..."
+        },
+        "Token": {
+          "keccak256": "0x456...",
+          "url": "https://raw.githubusercontent.com/ethereum/solidity/develop/std/Token.sol"
+        },
+        "mortal": {
+          "content": "contract mortal is owned { function kill() { if (msg.sender == owner) selfdestruct(owner); } }"
+        }
+      },
+      // This is a subset of the regular compiler input
+      // Its content is specific to the compiler (determined by the language and compiler fields)
+      settings:
+      {
+        remappings: [":g/dir"],
+        optimizer: {enabled: true, runs: 500},
+        compilationTarget: {
+          "myFile.sol": MyContract"
+        },
+        libraries: {
+          "MyLib": "0x123123..."
+        }
+      },
+      // This is a subset of the regular compiler output
+      output:
+      {
+        abi: [ /* abi definition */ ],
+        userdoc: [],
+        devdoc: [],
+      }
+    }
+
+This is used in the following way: A component that wants to interact
+with a contract (e.g. mist) retrieves the creation transaction of the contract
+and from that the first 33 bytes. If the first byte decodes into a PUSH32
```

Comment:
The whole metadata section was copied from this PR to #1386. Once that is merged, this one will only refer to it and won't copy the section.